runtime/src/main/java/org/corfudb/runtime/collections/UFOTxnBase.java

```diff
@@ -0,0 +1,369 @@
+     * @param value    Value or payload of the record.
+     * @param metadata Metadata associated with the record.
+     * @param forceSetMetadata if specified, metadata will not be validated, just set
+     * @param <K>      Type of Key.
+     * @param <V>      Type of Value.
+     * @param <M>      Type of Metadata.
+     * @return TxnContext instance.
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message>
+    UFOTxnBase putRecord(@Nonnull Table<K, V, M> table,
+                         @Nonnull final K key,
+                         @Nonnull final V value,
+                         @Nullable final M metadata,
+                         boolean forceSetMetadata) {
+
+        if (forceSetMetadata || metadata == null) {
+            this.txnContext = txnContext.put(table, key, value, null);
+            return this;
+        }
+
+        this.txnContext = txnContext.merge(table, key,
+                this::fixUpMetadata, new CorfuRecord<>(value, metadata));
+        return this;
+    }
+
+    /**
+     * Core logic for handling metadata modifications on mutations
+     * @param oldRecord - The previous fetched record that exists in the table.
+     * @param deltaUpdate - New record that will be inserted/updated.
+     * @param <K> - type of the key
+     * @param <V> - type of the value or payload
+     * @param <M> - type of the metadata
+     * @return the merged record that will be inserted into the table.
+     */
+    private <K extends Message, V extends Message, M extends Message>
+    CorfuRecord<V, M> fixUpMetadata(CorfuRecord<V, M> oldRecord, CorfuRecord<V, M> deltaUpdate) {
+        M deltaMetadata = deltaUpdate.getMetadata();
+            final Message.Builder builder = deltaMetadata.toBuilder();
+            long currentTime = System.currentTimeMillis();
+            for (Descriptors.FieldDescriptor fieldDescriptor : deltaMetadata.getDescriptorForType().getFields()) {
+                switch (fieldDescriptor.getName()) {
+                    case "revision":
+                        if (oldRecord == null || oldRecord.getMetadata() == null) {
+                            builder.setField(fieldDescriptor, 0L);
+                        } else {
+                            Long prevRevision = (Long) oldRecord.getMetadata().getField(fieldDescriptor);
+                            Long givenRevision = (Long)deltaMetadata.getField(fieldDescriptor);
+                            if ((givenRevision > 0 && // Validate revision only if set
+                                    prevRevision.longValue() == givenRevision.longValue())
+                            || givenRevision == 0) { // Do not validate revision if field isn't set
+                                builder.setField(fieldDescriptor, prevRevision + 1);
+                            } else {
+                                throw new StaleRevisionUpdateException(prevRevision, givenRevision);
+                            }
```

Comment:
Looks like revision == 0 means uninitialized (because on line 147 givenRevision == 0 means not set), but when creating the record for the first time (oldRecord == null), why do we set it to 0? Should it be 1?

Reply:
I realized but there is a bug that revision must start with 0!